### DISCUSSION TITLE
`ShellSpec` で半角英数字（`alphanumeric`）をパターンマッチさせたい

### DISCUSSION THREAD

**Discussion by KEINOS** [Qithub-BOT/QiiCipher#54] [Q&A] [answered]
## `ShellSpec` で n 桁の `alphanumeric` を検知したい

`ShellSpec` の文字列マッチで `a-f` `0-9` のハッシュ値であることを確認したいのですが、`^[a-f0-9]+$` のような正規表現が効きません。

`[a-f0-9]` の 1 桁、もしくは `[a-f0-9][a-f0-9]` と繰り返すと 2 桁は検知できるのですが、16 桁のハッシュとなると 16 回ぶん繰り返さないとならず、スマートではありません。

`[a-f0-9]{2}` や `[a-f0-9]+` のような繰り返しの指定もできないっぽいのですが、何かスマートな案はないでしょうか。

以下はパスしないサンプルのテストです。

```shell
#shellcheck shell=sh

sampleFunc() {
    printf "%s" 'abcdef1234567890' # 16 桁のハッシュ
}

Describe 'sampleFunc'
    It 'should return the version number of OpenSSH'
        When call sampleFunc

        The stdout should match pattern "[a-z0-9]{16}"
        The stdout should not include "\n"
    End
End
```

## 参考文献

> ## `match pattern` matcher
> 
> ```shell
> The output should match pattern <PATTERN>
> ```
>
> PATTERN examples
> 
> - `foo*`
> - `foo?`
> - `[fF]oo`
> - `[!F]oo`
> - `[a-z]`
> - `foo|bar`
>
> （[`match pattern` matcher](https://github.com/shellspec/shellspec/blob/master/docs/references.md#match-pattern-matcher) | string matchers | References | shellspec @ GitHub より）

**Comment by KEINOS** [comment]
詳しく咀嚼する前に（忘れる前に）メモ。... ... え？POSIX sh の限界？

> 表現 | 説明
> -- | --
> ? | 任意の1文字にマッチ
> | * | 任意の0文字以上にマッチ
> [abc] | 文字a,b,cいずれか1文字にマッチ
> [!abc] | 文字a,b,c以外の1文字にマッチ
> [a-z] | aからzまでの1文字にマッチ
> [!a-z] | aからz以外の1文字にマッチ
>
> 上の表を見てわかる通り、パターンマッチの能力はかなり貧弱です。特に、**直前の表現のn回繰り返し（正規表現で言う`*`）が使えない**のは非常に痛いですね。
>
> （ [パターンの書き方](https://qiita.com/hnw/items/291090f8632f3b40e1a0#%E3%83%91%E3%82%BF%E3%83%BC%E3%83%B3%E3%81%AE%E6%9B%B8%E3%81%8D%E6%96%B9) | POSIX shの変数展開による文字列置換の限界を探る @ Qiita より）

**Comment by yoshi389111** [comment, marked as answer]
わかりにくいですが grep を使えばできないわけではないと思います。

もっと良い方法もありそうですが。

```shell
#shellcheck shell=sh

sampleFunc() {
    printf "%s" 'abcdef1234567890' # 16 桁のハッシュ
}

isStdoutHex16() {
    # $1: stdout, $2: stderr, $3: status
    echo "$1" | grep -Eq '^[0-9a-f]{16}$'
    # ↑echo で改行コードがあっても大丈夫みたい
}

Describe 'sampleFunc'
    It 'should return the version number of OpenSSH'
        When call sampleFunc

        The result of "isStdoutHex16()" should be successful
        The stdout should not include "\n"
    End
End
```

* [function subject](https://github.com/shellspec/shellspec/blob/master/docs/references.md#function-subject)
* [result modifier](https://github.com/shellspec/shellspec/blob/master/docs/references.md#result-modifier)

**Reply by KEINOS** [reply]
先の「パターンの書き方」で「[特定の文字群が連続する限り取り出す](https://qiita.com/hnw/items/291090f8632f3b40e1a0#%E6%96%87%E5%AD%97%E5%88%97%E5%85%88%E9%A0%AD%E3%81%8B%E3%82%89%E7%89%B9%E5%AE%9A%E3%81%AE%E6%96%87%E5%AD%97%E7%BE%A4%E3%81%8C%E9%80%A3%E7%B6%9A%E3%81%99%E3%82%8B%E9%99%90%E3%82%8A%E5%8F%96%E3%82%8A%E5%87%BA%E3%81%99)」というのがありました。

これを使って「関数の出力結果の値」と「出力結果から特定の連続文字群を抜き出した値」が同じであれば、半角英数字かつ 16 進数小文字を検知できそうです。

```bash
result_raw="$(sampleFunc)"
result_alphanum="${result_raw%%[!a-f0-9]*}"

if [ "$result_raw" = "$result_alphanum" ]; then
    echo 'OK'
fi
```

正常系

```bash
#shellcheck shell=sh

sampleFuncOK() {
    printf "%s" 'abcdef1234567890' # alphanumeric
}

Describe 'sampleFuncOK'
    It 'should be success'
        result_raw="$(sampleFuncOK)"
        result_alphanum="${result_raw%%[!a-f0-9]*}"

        The variable result_raw should equal "$result_alphanum"
        The variable result_raw should not include "\n"
    End
End
```

異常系

```bash
sampleFuncNG() {
    printf "%s" 'abcdef123456789!' # non alphanumeric
}

Describe 'sampleFuncNG'
    It 'should not be success'
        result_raw="$(sampleFuncNG)"
        result_alphanum="${result_raw%%[!a-f0-9]*}"

        The variable result_raw should not equal "$result_alphanum"
        The variable result_raw should not include "\n"
    End
End
```

... でも `grep` のヘルパー関数を作って、それを使った方が直感的でいいっすね。そうしよーっと。

**Reply by KEINOS** [reply]
（シェルでもヘルパー関数っていうのか？）

テスト用に `grep` を使った関数を作る方向にします。`isStdoutHex16` でベストアンサー！ 👌
